interfaces/apparmor/backend_test.go

```diff
@@ -492,7 +495,362 @@ func (s *backendSuite) TestSetupHostSnapConfineApparmorForReexecWritesNew(c *C)
+	// NOTE: Errors in determining NFS are non-fatal to prevent snapd from
+	// failing to operate. A warning message is logged but system operates as
+	// if NFS was not active.
+	c.Assert(err, IsNil)
+
+	// While other stuff failed we created the policy directory and didn't
+	// write any files to it.
+	files, err := ioutil.ReadDir(dirs.SnapConfineAppArmorDir)
+	c.Assert(err, IsNil)
+	c.Assert(files, HasLen, 0)
+
+	// We didn't reload the policy.
+	c.Assert(cmd.Calls(), HasLen, 0)
+}
+
+// Test behavior when os.Readlink "/proc/self/exe" fails.
+func (s *backendSuite) TestSetupSnapConfineGeneratedPolicyError2(c *C) {
+	// Make it appear as if NFS was mounted under /home.
+	restore := apparmor.MockMountInfo("680 27 0:59 / /home/zyga/nfs rw,relatime shared:478 - nfs4 localhost:/srv/nfs rw,vers=4.2,rsize=524288,wsize=524288,namlen=255,hard,proto=tcp,port=0,timeo=600,retrans=2,sec=sys,clientaddr=127.0.0.1,local_lock=none,addr=127.0.0.1")
+	defer restore()
+	restore = apparmor.MockEtcFstab("")
+	defer restore()
+
+	// Intercept interaction with apparmor_parser
+	cmd := testutil.MockCommand(c, "apparmor_parser", "")
+	defer cmd.Restore()
+
+	// Intercept the /proc/self/exe symlink and point it to the snapd from the
+	// mounted core snap. This indicates that snapd has re-executed and
+	// should not reload snap-confine policy.
+	fakeExe := filepath.Join(s.RootDir, "corrupt-proc-self-exe")
+	restore = apparmor.MockProcSelfExe(fakeExe)
+	defer restore()
+
+	// Setup generated policy for snap-confine.
+	err := apparmor.SetupSnapConfineGeneratedPolicy()
+	c.Assert(err, ErrorMatches, "cannot read /proc/self/exe,.*")
+
+	// We created the policy file.
+	files, err := ioutil.ReadDir(dirs.SnapConfineAppArmorDir)
+	c.Assert(err, IsNil)
+	c.Assert(files, HasLen, 1)
+
+	// We didn't reload the policy though.
+	c.Assert(cmd.Calls(), HasLen, 0)
+}
+
+// Test behavior when exec.Command "apparmor_parser" fails
+func (s *backendSuite) TestSetupSnapConfineGeneratedPolicyError3(c *C) {
+	// Make it appear as if NFS was mounted under /home.
+	restore := apparmor.MockMountInfo("680 27 0:59 / /home/zyga/nfs rw,relatime shared:478 - nfs4 localhost:/srv/nfs rw,vers=4.2,rsize=524288,wsize=524288,namlen=255,hard,proto=tcp,port=0,timeo=600,retrans=2,sec=sys,clientaddr=127.0.0.1,local_lock=none,addr=127.0.0.1")
+	defer restore()
+	restore = apparmor.MockEtcFstab("")
+	defer restore()
+
+	// Intercept interaction with apparmor_parser and make it fail.
+	cmd := testutil.MockCommand(c, "apparmor_parser", "echo testing; exit 1")
+	defer cmd.Restore()
+
+	// Intercept the /proc/self/exe symlink.
+	fakeExe := filepath.Join(s.RootDir, "fake-proc-self-exe")
+	err := os.Symlink("/usr/lib/snapd/snapd", fakeExe)
+	c.Assert(err, IsNil)
+	restore = apparmor.MockProcSelfExe(fakeExe)
+	defer restore()
+
+	// Setup generated policy for snap-confine.
+	err = apparmor.SetupSnapConfineGeneratedPolicy()
+	c.Assert(err, ErrorMatches, "cannot reload snap-confine apparmor profile: testing")
```

Comment:
If the policy load fails, I seems to make sense to clean up the generated profile?

Reply:
Hmmm, I think the re-exec-from-core case may still want that but if we cannot load it ourselves then yeah, lets remove it. I'll make the adjustment.

Reply:
Done.